csrc/trtllm_fmha_kernel_launcher.cu

```diff
@@ -25,317 +28,194 @@
-
-  runner_params.qPtr = q_heads;
-
-  runner_params.mQkvLayout = QkvLayout::PagedKv;
-  runner_params.kvPtr = cache_heads;
-  runner_params.kvPageIdxPtr = block_tables.data_ptr<KVCachePageIndex>();
-  runner_params.mMaxNumPagesPerSeqKv = max_num_blocks_per_seq;
-  runner_params.mNumTokensPerPage = tokens_per_page;
-
-  runner_params.seqLensKvPtr = reinterpret_cast<int const*>(seq_lens.data_ptr<int>());
-
-  runner_params.oPtr = output_ptr;
-  runner_params.mHeadDimQk = head_size;
-  runner_params.mHeadDimV = head_size;
-  runner_params.mNumHeadsQ = num_heads;
-  runner_params.mNumHeadsKv = num_k_heads;
-  runner_params.mNumHeadsQPerKv = num_heads / num_k_heads;
-  runner_params.mBatchSize = batch_size;
-  runner_params.mMaxSeqLenQ = query.size(0);
-  runner_params.mMaxSeqLenKv = max_seq_len;
-
-  runner_params.mSumOfSeqLensQ = sum_seq_q;
-  runner_params.mSumOfSeqLensKv = sum_seq_kv;
-  runner_params.cumSeqLensQPtr = cum_seq_lens_q.data_ptr<int>();
-  runner_params.cumSeqLensKvPtr = cum_seq_lens_kv.data_ptr<int>();
 
-  runner_params.mScaleQ = 1.0;
-  // Set the chunked attention size and sliding window size to INT_MAX to disable them when checking
-  // if
-  // the kernel is supported.
-  runner_params.mChunkedAttentionSize = INT_MAX;
-  runner_params.mAttentionWindowSize = window_left == -1 ? INT_MAX : window_left + 1;
-
-  runner_params.mMultiProcessorCount = getMultiProcessorCount();
-  auto const [free_memory, total_memory] = getDeviceMemoryInfo(false);
   auto [foundKernels, kinfo] = fmha_runner.isSupportedWithInfo(runner_params);
   if (!foundKernels) {
     std::ostringstream err_msg;
-    TllmGenSelectKernelParams select_kernel_params{runner_params};
-    err_msg << "Missing TRTLLM-GEN context kernel:" << kinfo;
+    err_msg << "Missing TRTLLM-GEN kernel ("
+            << (mode == TllmPagedAttentionMode::Context ? "context" : "decode") << "): " << kinfo;
     FLASHINFER_ERROR(err_msg.str());
   }
-  int max_head_dim_kv = head_size;
-
-  runner_params.mNumPagesInMemPool =
-      total_memory / (runner_params.mNumHeadsKv * runner_params.mNumTokensPerPage *
-                      max_head_dim_kv * get_size_in_bytes(CACHE_T));
 
-  runner_params.outputScale = bmm2_scale;
-  runner_params.scaleSoftmaxLog2 = bmm1_scale * M_LOG2E;
-
-  runner_params.stream = stream;
   fmha_runner.run(runner_params);
 }
 
+void trtllm_paged_attention_decode(at::Tensor& out, at::Tensor& query, at::Tensor& key_value_cache,
+                                   at::Tensor& workspace_buffer, at::Tensor& block_tables,
+                                   at::Tensor& seq_lens, int64_t max_kv_len, double bmm1_scale,
+                                   double bmm2_scale, int64_t window_left, int64_t sm_count) {
+  DISPATCH_PYTORCH_DTYPE_TO_CTYPE(query.scalar_type(), DTypeQ, [&]() {
+    return DISPATCH_PYTORCH_DTYPE_TO_CTYPE(key_value_cache.scalar_type(), DTypeKV, [&]() {
+      return DISPATCH_PYTORCH_DTYPE_TO_CTYPE(out.scalar_type(), DTypeO, [&]() {
```

Comment:
P1 (not a must in this PR, but would be great in next PR): we should check if the provided arguments are as expected. If something is not expected, raise an error to let callers know that they did it wrong.

The checks should include but not limited to:

- Check the tensors': dtype, rank, dimensions, device, is_contiguous, etc.
- Check if all tensors are on the same device.
- Check if the dtype combinations are supported.